### ISSUE TITLE
Unexpected behavior when using multiple signing keys for an identity zone

### ISSUE THREAD

**Issue by mbalsam82** [cloudfoundry/uaa#1440]
**SECURITY NOTICE: If you have found a security problem in the UAA, please do not file a public github issue. Instead, please send an email to security@cloudfoundry.org**

Thanks for taking the time to file an issue. You'll minimize back and forth and help us help you more effectively by answering all of the following questions as specifically and completely as you can.

### What version of UAA are you running?

develop branch (current release 74.27.0)

### How are you deploying the UAA?

I am deploying the UAA

- locally only using gradlew

### What did you do?

I tried both, to create and update an identity zone with the following token policy that contains two signing keys for **seamless signing key rotation**:
```
...
"config": {
    ...
    "tokenPolicy" : {
      "accessTokenValidity" : 3600,
      "refreshTokenValidity" : 7200,
      "jwtRevocable" : false,
      "refreshTokenUnique" : false,
      "refreshTokenFormat" : "opaque",
      "activeKeyId" : "key-2",
      "keys" : {
          "key-1" : {
            "signingKey": "some.key"
          },
          "key-2" : {
            "signingKey": "some.other.key"
          }
      }
}
```

### What did you expect to see? What goal are you trying to achieve with the UAA?

The identity zone should have the expected token policy config:
```
"config": {
    ...
    "tokenPolicy" : {
      "accessTokenValidity" : 3600,
      "refreshTokenValidity" : 7200,
      "jwtRevocable" : false,
      "refreshTokenUnique" : false,
      "refreshTokenFormat" : "opaque",
      "activeKeyId" : "key-2"
    }
}
```
Actually, I would also expect to see the list of keys, or at least a list of their IDs.

In addition, the `/token_keys` response for the identity zone should contain the public keys for both signing keys.

### What did you see instead?

Instead, the identity zone has defaults for all fields in the token policy and no active key:
```
"config": {
    ...
    "tokenPolicy" : {
      "accessTokenValidity" : -1,
      "refreshTokenValidity" : -1,
      "jwtRevocable" : false,
      "refreshTokenUnique" : false,
      "refreshTokenFormat" : "jwt",
      "activeKeyId" : null
    }
}
```
The `/token_keys` response for the identity zone contains only the _legacy-signing-key_.

The logs only show a successful IdentityZoneModifiedCreated/IdentityZoneModifiedEvent.

**Comment by mbalsam82** [comment]
This makes a rolling signing key rotation, where tokens signed with the old key can still be validated, impossible.

**Comment by mbalsam82** [comment]
I debugged the case with 2 keys and it looks like it's a parsing issue. The keys are already missing in the `IdentityZone` request body param for `IdentityZoneEndpoints#createIdentityZone`.

**Comment by mbalsam82** [comment]
This can be found in the API reference under Token Key(s):

> JWT signing keys are specified via the identity zone configuration (see /identity-zones). An identity zone token policy can be configured with multiple keys for purposes of key rotation. When adding a new key, set its ID as the activeKeyId to use it to sign all new tokens. /introspect will continue to verify tokens signed with the previous signing key for as long as it is present in the keys of the identity zone's token policy. Remove it to invalidate all those tokens.

Besides not being able to configure an identity zone token policy with multiple singing keys, I'm really wondering how this is supposed to work.
- All there is to modify an identity zone is a PUT request while the GET request removes the signing keys from the response, i.e. I'd have to redundantly store all signing keys outside of UAA to be able to add a key to the list or remove one.
- There are no endpoints to add or remove a signing key to an identity zone - or am I missing something?

**Comment by mbalsam82** [comment]
I've been able to successfully create a second signing key against a remote deployment that uses an earlier UAA version (74.18.0).

**Comment by strehle** [comment]
HI @mbalsam82 ,
tried this but could not reproduce it.
Do you use HMAC or RSA signing key ? I tried with 2nd RSA

**Comment by mbalsam82** [comment]
> HI @mbalsam82 ,
> tried this but could not reproduce it.
> Do you use HMAC or RSA signing key ? I tried with 2nd RSA

I used RSA signing keys iirc. I created this bug a while ago when I noticed it on the develop branch. It hasn't occurred again with the latest releases, so I'd say it's **obsolete** and I'll close the issue. I still think that UAA should have better APIs to be able to rotate signing keys in a meaningful way (i.e. without having to keep them somewhere outside of UAA), but that'd be a separate issue.
